doc/HelpInterface.md

```diff
@@ -0,0 +1,89 @@
+
+someuser> !help quote !quoteadd
+halibot1> Add a new quote to the db
+halibot1> Call like so: !quoteadd text of new quote here
+
+```
+
+### Via command line
+
+*do we actually want to support this?*
+
+## How to implement help text?
+
+```python
+class Quote(HalModule):
+    ...
+    def help(args=None):
+        if args == "!quoteadd":
+              return "Add a new quote to the db\nCall like so: !quoteadd text of new quote here"
+        return "Display quotes from a file. Available commands are !quote, !quoteadd, !quotedel"
+```
+
+That's it!
+Nothing else needs to be done to support text through the !help interface.
+
+
+# Specification
+
+## Help Module
+
+The Help module is fairly simple at its core.
+It handles commands of the form `!help [module] [module topic]`
+
+If nothing is supplied, it responds with a generic greeting, instructing the user to call again with `!help [module]` to learn more about specific modules.
+It also should return a list of the currently loaded modules (which should also be available via the `admin` module's `!modules` command).
+
+When only `module` is supplied (e.g. `!help quote`), the help module attempts to call the `help()` method on the selected module.
+So in the example of the `quote` module, `!help quote` will find the module named `quote`, and call `.help()` on it via the `.invoke()` feature (see ModuleInteractivity.md for more on `.invoke()`).
+If the specified module does not implement a `.help()` method, an apologetic error message is return instead.
+
+If there are more than two words (separated by spaces) supplied to `!help`, the first word is considered to be module to inquire about, and everything after that space is considered supplementary text to query the module.
+This supplementary text is passed into the module's `.help()` method, where it may or may not decide to respond differently.
+Again, using the above example, `!help quote !quoteadd` invokes `quote`'s `.help()` method with the argument `"!quoteadd"`.
+See the next section for recommendations on handling help topics.
+
+## The `.help()` method
+
+Help text on a module at its core, is just a function named `help` that returns a string.
+All arguments to `.help()` should be treated as a string, and `.help()` should ONLY return a string.
+Arguments may be omitted (as in, nonexistant, not just passing `None` or `""`), and thus the method should be able to handle either case.
+Arguments do not need to be handled; the help method may choose to ignore any arguments.
+Returned help text does not need to be deterministic, help methods may choose to modify the text based on the current running state.
+
+## Tips for implementing a `.help()`.
+
+### Output Formatting
+
+As a general rule, try to avoid help text longer than three lines.
+Most users will likely be coming from some kind of chat protocol like IRC or XMPP, therefore long messages could trigger a server rate limit.
```

Comment:
This is also probably a good use-case for protocol specific formatters, and format checkers. XMPP can handle large blobs of text pretty okay (although is a bit spammy), but IRC doesn't like it.

Also, if this is in a PM, longer text could also be possibly supplied, as it wouldn't flood a channel.